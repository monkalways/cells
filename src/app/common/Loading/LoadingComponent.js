import React from 'react';
import PropTypes from 'prop-types';
import { CircularProgress, withStyles } from '@material-ui/core';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  size: PropTypes.number,
};

const defaultProps = {
  size: 100,
};

const LoadingComponent = ({ classes, size }) => (
  <div className={classes.container}>
    <CircularProgress className={classes.progress} size={size} />
  </div>
);

LoadingComponent.propTypes = propTypes;
LoadingComponent.defaultProps = defaultProps;

export default withStyles((theme) => ({
  container: {
    display: 'flex',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  progress: {
    margin: theme.spacing.unit * 2,
  },
}))(LoadingComponent);
