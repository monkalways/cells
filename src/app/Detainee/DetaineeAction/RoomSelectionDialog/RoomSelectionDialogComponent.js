import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormControl,
  FormControlLabel,
  FormGroup,
  Grid,
  Radio,
  RadioGroup,
  withStyles,
  Typography,
} from '@material-ui/core';

import Loading from '../../../common/Loading';

const propTypes = {
  areActivityRoomsRefreshing: PropTypes.bool.isRequired,
  availableRooms: PropTypes.arrayOf(PropTypes.string),
  classes: PropTypes.shape({}).isRequired,
  currentRoom: PropTypes.string,
  detainee: PropTypes.shape({
    firstName: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
  }).isRequired,
  getAvailableActivityRoomsRefresh: PropTypes.func.isRequired,
  isAnyRoomForGivenActivityAvailable: PropTypes.bool.isRequired,
  isDialogOpen: PropTypes.bool.isRequired,
  isAssigningToRoom: PropTypes.bool.isRequired,
  moveDetaineeToActivityRoom: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  usage: PropTypes.string.isRequired,
};

const defaultProps = {
  availableRooms: [],
  currentRoom: null,
};

export class RoomSelectionDialogComponent extends Component {
  state = {
    value: null,
  };

  componentWillReceiveProps(nextProps) {
    const { value } = this.state;
    if (!value && nextProps.availableRooms) {
      this.setState({
        value: nextProps.areActivityRoomsRefreshing
          ? null
          : nextProps.availableRooms[0],
      });
    }
  }

  handleConfirmClick = () => {
    const { detainee, currentRoom, moveDetaineeToActivityRoom } = this.props;
    const { value } = this.state;

    moveDetaineeToActivityRoom(detainee.id, currentRoom, value);
  };

  handleOpen = () => {
    const { getAvailableActivityRoomsRefresh } = this.props;
    getAvailableActivityRoomsRefresh();
  };

  handleChange = (event) => {
    this.setState({ value: event.target.value });
  };

  render() {
    const {
      areActivityRoomsRefreshing,
      availableRooms,
      classes,
      detainee,
      isAnyRoomForGivenActivityAvailable,
      isAssigningToRoom,
      isDialogOpen,
      onClose,
      usage,
    } = this.props;

    const { value } = this.state;

    return (
      <Dialog
        disableBackdropClick={isAssigningToRoom}
        disableEscapeKeyDown={isAssigningToRoom}
        open={isDialogOpen}
        onClose={onClose}
        onRendered={this.handleOpen}
      >
        {areActivityRoomsRefreshing ? (
          <React.Fragment>
            <DialogTitle>Searching for available room...</DialogTitle>
            <DialogContent>
              <Loading size={50} />
            </DialogContent>
          </React.Fragment>
        ) : isAnyRoomForGivenActivityAvailable ? (
          <React.Fragment>
            <DialogTitle>
              {`Please select ${usage} room for ${detainee.lastName}, ${
                detainee.firstName
              }:`}
            </DialogTitle>
            <DialogContent>
              {isAssigningToRoom ? (
                <Loading size={50} />
              ) : (
                <Grid container justify="center" alignItems="center">
                  <FormControl required component="fieldset">
                    <FormGroup>
                      <RadioGroup
                        name="availableRooms"
                        value={value}
                        onChange={this.handleChange}
                      >
                        {availableRooms.map((room) => (
                          <FormControlLabel
                            key={room}
                            value={room}
                            control={<Radio className={classes.radioButton} />}
                            label={<Typography variant="h6">{room}</Typography>}
                            disabled={isAssigningToRoom}
                          />
                        ))}
                      </RadioGroup>
                    </FormGroup>
                  </FormControl>
                </Grid>
              )}
            </DialogContent>
            <DialogActions>
              <Button
                id="cancelButton"
                onClick={onClose}
                color="primary"
                disabled={isAssigningToRoom}
              >
                Cancel
              </Button>
              <Button
                id="confirmButton"
                onClick={this.handleConfirmClick}
                color="primary"
                variant="contained"
                autoFocus
                disabled={isAssigningToRoom}
              >
                Confirm
              </Button>
            </DialogActions>
          </React.Fragment>
        ) : (
          <React.Fragment>
            <DialogTitle>Sorry, no rooms are currently available.</DialogTitle>
            <DialogActions>
              <Button id="closeButton" onClick={onClose} color="primary">
                Close
              </Button>
            </DialogActions>
          </React.Fragment>
        )}
      </Dialog>
    );
  }
}

RoomSelectionDialogComponent.propTypes = propTypes;
RoomSelectionDialogComponent.defaultProps = defaultProps;

export default withStyles((theme) => ({
  radioButton: {
    padding: theme.spacing.unit,
  },
}))(RoomSelectionDialogComponent);
