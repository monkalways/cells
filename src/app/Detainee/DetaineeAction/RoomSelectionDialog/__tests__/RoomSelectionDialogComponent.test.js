import React from 'react';
import { shallow } from 'enzyme';
import { Dialog, RadioGroup } from '@material-ui/core';

import { RoomSelectionDialogComponent } from '../RoomSelectionDialogComponent';
import Loading from '../../../../common/Loading';

describe('RoomSelectionDialogComponent', () => {
  let areActivityRoomsRefreshing;
  let availableRooms;
  let currentRoom;
  let destinationRoom;
  let detainee;
  let getAvailableActivityRoomsRefresh;
  let isAnyRoomForGivenActivityAvailable;
  let isDialogOpen;
  let isAssigningToRoom;
  let moveDetaineeToActivityRoom;
  let onClose;
  let usage;

  beforeEach(() => {
    areActivityRoomsRefreshing = false;
    availableRooms = [];
    currentRoom = 'B4';
    destinationRoom = 'T4';
    detainee = {
      firstName: 'Bob',
      id: '123',
      lastName: 'Saget',
    };
    getAvailableActivityRoomsRefresh = jest.fn();
    isAnyRoomForGivenActivityAvailable = false;
    isDialogOpen = false;
    isAssigningToRoom = false;
    moveDetaineeToActivityRoom = jest.fn();
    onClose = jest.fn();
    usage = 'usage';
  });

  const setup = () => {
    const classes = { radioButton: 'radioButton' };

    return shallow(<RoomSelectionDialogComponent
      areActivityRoomsRefreshing={areActivityRoomsRefreshing}
      availableRooms={availableRooms}
      classes={classes}
      currentRoom={currentRoom}
      destinationRoom={destinationRoom}
      detainee={detainee}
      getAvailableActivityRoomsRefresh={getAvailableActivityRoomsRefresh}
      isAnyRoomForGivenActivityAvailable={isAnyRoomForGivenActivityAvailable}
      isDialogOpen={isDialogOpen}
      isAssigningToRoom={isAssigningToRoom}
      moveDetaineeToActivityRoom={moveDetaineeToActivityRoom}
      onClose={onClose}
      usage={usage}
    />);
  };

  it('should render Loading component if activity rooms are refreshing', () => {
    areActivityRoomsRefreshing = true;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Loading)).toExist();
  });

  it('should show a close button when the room is no longer available', () => {
    areActivityRoomsRefreshing = false;
    isAnyRoomForGivenActivityAvailable = false;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#closeButton')).toExist();
  });

  it('should show confirm and cancel buttons when moving a detainee', () => {
    areActivityRoomsRefreshing = false;
    isAnyRoomForGivenActivityAvailable = true;
    isAssigningToRoom = false;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#cancelButton')).toExist();
    expect(wrapper.find('#confirmButton')).toExist();
    expect(wrapper.find(Loading)).not.toExist();
  });

  it('should show a loading component when it is assigning the detainee to a room', () => {
    areActivityRoomsRefreshing = false;
    isAnyRoomForGivenActivityAvailable = true;
    isAssigningToRoom = true;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#cancelButton')).toExist();
    expect(wrapper.find('#confirmButton')).toExist();
    expect(wrapper.find(Loading)).toExist();
  });

  it('should handle the dialog closing', () => {
    const wrapper = setup();
    wrapper.find(Dialog).simulate('close');
    expect(onClose).toBeCalled();
  });

  it('should handle close button click', () => {
    areActivityRoomsRefreshing = false;
    isAnyRoomForGivenActivityAvailable = false;

    const wrapper = setup();
    wrapper.find('#closeButton').simulate('click');
    expect(onClose).toBeCalled();
  });

  it('should handle confirm button click', () => {
    areActivityRoomsRefreshing = false;
    isAnyRoomForGivenActivityAvailable = true;

    const wrapper = setup();
    wrapper.setState({ value: 'T4' });
    wrapper.find('#confirmButton').simulate('click');
    expect(moveDetaineeToActivityRoom).toBeCalledWith(
      detainee.id,
      currentRoom,
      'T4',
    );
  });

  it.each([
    [null, { areActivityRoomsRefreshing: true, availableRooms: null }, null],
    [
      null,
      { areActivityRoomsRefreshing: false, availableRooms: ['T1', 'T2'] },
      'T1',
    ],
    ['T1', { areActivityRoomsRefreshing: false, availableRooms: null }, 'T1'],
  ])(
    'should set value to first available room when activity rooms have finished refreshing',
    (previousValue, nextProps, nextValue) => {
      const wrapper = setup();
      wrapper.setState({ value: previousValue });
      wrapper.instance().componentWillReceiveProps(nextProps);
      expect(wrapper).toHaveState('value', nextValue);
    },
  );

  it('should update value when radio selection changes', () => {
    areActivityRoomsRefreshing = false;
    isAnyRoomForGivenActivityAvailable = true;
    isAssigningToRoom = false;

    const wrapper = setup();
    const event = {
      target: {
        value: 'CHU1',
      },
    };
    wrapper.find(RadioGroup).simulate('change', event);
    expect(wrapper.state().value).toEqual('CHU1');
  });

  it('should handle cancel button click', () => {
    areActivityRoomsRefreshing = false;
    isAnyRoomForGivenActivityAvailable = true;

    const wrapper = setup();
    wrapper.find('#cancelButton').simulate('click');
    expect(onClose).toBeCalled();
  });
});
