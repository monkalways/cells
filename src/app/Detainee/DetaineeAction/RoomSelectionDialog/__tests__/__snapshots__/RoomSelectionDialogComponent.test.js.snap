// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomSelectionDialogComponent should render Loading component if activity rooms are refreshing 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={false}
  disableEscapeKeyDown={false}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Searching for available room...
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <WithStyles(LoadingComponent)
      size={50}
    />
  </WithStyles(DialogContent)>
</WithStyles(Dialog)>
`;

exports[`RoomSelectionDialogComponent should show a close button when the room is no longer available 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={false}
  disableEscapeKeyDown={false}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Sorry, no rooms are currently available.
  </WithStyles(DialogTitle)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      id="closeButton"
      onClick={[MockFunction]}
    >
      Close
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`RoomSelectionDialogComponent should show a loading component when it is assigning the detainee to a room 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Please select usage room for Saget, Bob:
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <WithStyles(LoadingComponent)
      size={50}
    />
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      disabled={true}
      id="cancelButton"
      onClick={[MockFunction]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      autoFocus={true}
      color="primary"
      disabled={true}
      id="confirmButton"
      onClick={[Function]}
      variant="contained"
    >
      Confirm
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`RoomSelectionDialogComponent should show confirm and cancel buttons when moving a detainee 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={false}
  disableEscapeKeyDown={false}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Please select usage room for Saget, Bob:
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <WithStyles(Grid)
      alignItems="center"
      container={true}
      justify="center"
    >
      <WithStyles(FormControl)
        component="fieldset"
        required={true}
      >
        <WithStyles(FormGroup)>
          <RadioGroup
            name="availableRooms"
            onChange={[Function]}
            value={null}
          />
        </WithStyles(FormGroup)>
      </WithStyles(FormControl)>
    </WithStyles(Grid)>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      disabled={false}
      id="cancelButton"
      onClick={[MockFunction]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      autoFocus={true}
      color="primary"
      disabled={false}
      id="confirmButton"
      onClick={[Function]}
      variant="contained"
    >
      Confirm
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
